### DISCUSSION TITLE
[Feature Request] Accessing Versal BEL Attributes in RapidWright

### DISCUSSION THREAD

**Discussion by nqdtan** [Xilinx/RapidWright#598] [General]
In Versal devices, I find that there are some specific BELs whose attributes appear to be read-only to users in Vivado, and could only be set internally by the tool through some optimization phases. Examples are those of IMUX registers, GCLK_DELAY cells, etc. They help the design achieve better timing as far as I understood. The BEL attribute values are stored in a `*.belattrdb` file in a DCP file. If this file is removed from the DCP, Vivado reports a worsen timing slack. Attached is an example of the same design with two different DCPs (one with and the other without `belattrdb`), and their respective timing reports. We could observe that the one with `belattrdb` gets better timing result.

While RapidWright does preserve the Versal BEL attribute file in DCP, I'd imagine that one may attempt to relocate their design netlist to a different region, thus render the original BEL attributes invalid. Another case is combining multiple DCPs with different `belattrdb` files. There currently seems to be no possible way to modify or update the BEL attributes in accordance with the relocated design netlist. While this issue does not impact the functionality, I'm quite concerned with the negative impact on the performance (achieving lower Fmax) of the relocated design.

I think it would be helpful if there is some API in RapidWright that would allow us to parse and modify `belattrdb`.

[belattrdb_issue](https://drive.google.com/file/d/1IiNrvTvUvaT5v_uEhvr46UkRFN6dypgM/view?usp=sharing)

**Comment by clavin-xlnx** [comment]
This will take some thought and time, I will try to provide an update later in the week on how we might be able to proceed here.

**Reply by clavin-xlnx** [reply]
We have a way to address this issue.  We'll publish a solution soon.

**Reply by nqdtan** [reply]
Thank you for the update!

**Reply by clavin-xlnx** [reply]
@nqdtan - some basic APIs have been added into the new release branch `2022.2.1`, in the `Design` class:
- `public Map<Site, SiteConfig> getBELAttrs()`, where `SiteConfig` is a [new class](https://github.com/Xilinx/RapidWright/blob/96806abe0f134fba9cc2a769ccd11ec8693f249f/src/com/xilinx/rapidwright/design/SiteConfig.java#L49).
- `public BELAttr addBELAttr(Net net, Site site, SiteTypeEnum type, BEL bel, String name, String value)` , where `BELAttr` is a [new class](https://github.com/Xilinx/RapidWright/blob/96806abe0f134fba9cc2a769ccd11ec8693f249f/src/com/xilinx/rapidwright/design/BELAttr.java#L29).

The map returned from the first is mutable enables updates and changes.  In a future update, we can add the ability to automatically make updates for relocation purposes when we have access to a set of examples that require it.

**Reply by nqdtan** [reply]
This is working well for my case. Thank you very much for your swift response to include this in the upcoming release!
